c_src/ep_decoder.c

```diff
@@ -896,12 +896,19 @@ fill_default(ErlNifEnv *env, ep_spot_t *spot)
 
             case field_float:
             case field_double:
-                *t++ = state->double_zero;
+                *t++ = enif_make_double(env, 0.0);
                 break;
 
             case field_string:
             case field_bytes:
-                *t++ = state->binary_nil;
+                {
```

Comment:
Have you chosen to be safe by allocating an initialized bin? Before it was a nil binary (state->binary_nil)?

Reply:
yes, i think binaries are also garbage-collected similar to numbers so we have to allocate them on the spot. on line 935 murali explained a similar problem.

Reply:
and yes, just the change to doubles alone made the decode function stop causing segfaults, but i wanted to cover everything here to be safe

Reply:
why is using binary_nil a problem?

Reply:
after it's allocated when the state object is initalized, i think it gets garbage-collected by the erlang vm before we try to decode any protos. when we try to use it again, it's a hanging reference. 
this likely isn't a problem for atoms because their internal representation is different.

Reply:
Many Thanks for all the answers, Thomas. My understanding is that the fix in this PR is short term. We have to figure out why state is becoming corrupt and fix its cause.

Reply:
i forgot to mention that the upstream fork has also made this change - they've entirely removed the fields which went bad from the object and replaced them with a change very similar to this. there probably is a way to store these objects properly in the state, but it might be better to just go along with the changes upstream.